daliuge-engine/dlg/apps/pyfunc.py

```diff
@@ -712,6 +712,19 @@ def write_results(self, result):
                 elif self.output_parser is DropParser.NPY:
                     drop_loaders.save_npy(o, r)
                 elif self.output_parser is DropParser.RAW:
+                    if not isinstance(r, (bytes, memoryview, str)):
+                        raise Exception(
+                            "Can't write data of this type: ", type(r).__name__
+                        )
+                    o.write(r)
+                elif self.output_parser is DropParser.UTF8:
+                    if not isinstance(r, (bytes, memoryview, str)):
+                        try:
+                            r = str(r)
+                        except:  # not sure what the exception would be...
```

Comment:
It's quite unlikely that this lead to a exception/error, but my money would be a `ValueError`; this is the parent exception of `UnicodeError` which is probably the most likely error to be raised here. 

A secondary, catch-all error I would raise here might be `RuntimeError`.